src/estimators/stratified_sketch.py

```diff
@@ -0,0 +1,242 @@
+
+    Args:
+      set_generator: SetGenerator object to draw ids from for initialization.
+      CardinalitySketch: Class type of cardinality sketches this stratified
+        sketch will hold.
+
+    """
+
+    exact_multi_set = ExactMultiSet()
+    for set in set_generator:
+        for item in set:
+            exact_multi_set.add(item)
+    self.init_from_exact_multi_set(exact_multi_set, CardinalitySketch)
+
+  def assert_compatible(self, other):
+    """"Check if the two StratifiedSketch are comparable.
+
+    Args:
+      other: the other StratifiedSketch for comparison.
+
+    Raises:
+      AssertionError: if the other sketches are not StratifiedSketch, or if
+      their random_seed are different, or if the frequency targets are
+      different.
+    """
+    assert isinstance(other, StratifiedSketch), (
+        'other is not a StratifiedSketch.')
+    assert self.seed == other.seed, (
+        'The random seeds are not the same: '
+        f'{self.seed} != {other.seed}')
+    assert self.max_freq == other.max_freq, (
+        'The frequency targets are different: '
+        f'{self.max_freq} != {other.max_freq}')
+
+class PairwiseEstimator(object):
+  """Merge and estimate two StratifiedSketch."""
+
+  def __init__(self, sketch_operator):
+    """Create an estimator for two Stratified sketches.
+
+    Args:
+      sketch_operator: an object that have union, intersection, and
+        difference methods for two sketches.
+    """
+
+    self.sketch_union = sketch_operator.union
+    self.sketch_difference = sketch_operator.difference
+    self.sketch_intersection = sketch_operator.intersection
+
+  def __call__(self, this, that):
+    merged = self.merge_sketches(this, that)
+    return merged
+
+  def merge_sketches(self, this, that):
+    """Merge two StratifiedSketch.
+
+       Given 2 sketches A and B:
+       Merged(k) = (A(k) & B(0)) U (A(k-1) & B(1)) ... U (A(0) & B(k))
+         where
+         A(k) & B(0) =  A(k) - (A(k) & B(1+))
+         B(k) & A(0) =  B(k) - (B(k) & A(1+))
+    Args:
+      this: one of the two StratifiedSketch to be merged.
+      that: the other StratifiedSketch to be merged.
+
+    Returns:
+      A merged StratifiedSketch from the input.
+    """
+    assert isinstance(this, StratifiedSketch)
+    this.assert_compatible(that)
+    max_freq = this.max_freq
+    merged_sketch = copy.deepcopy(this)
+
+    for k in range(max_freq - 1):
+      # Calculate A(k) & B(0) = A(k) - (A(k) & B(1+))
+      merged = self.sketch_difference(
+          this.sketches[str(k + 1)], self.sketch_intersection(
+              this.sketches[str(k + 1)], that.sketches[ONE_PLUS]))
+
+      # Calculate A(0) & B(k) = B(k) - (B(k) & A(1+))
+      merged = self.sketch_union(
+          merged, self.sketch_difference(
+              that.sketches[str(k + 1)], self.sketch_intersection(
+                  this.sketches[ONE_PLUS], that.sketches[str(k + 1)])))
+
+      # Calculate A(i) & B(k-i)
+      for i in range(k):
```

Comment:
Here, it would be clearer to use range(1,k+1)

Reply:
actually, I modified it. to range(1, k) and changed the code below a little. I think this is cleaner, what do you think?